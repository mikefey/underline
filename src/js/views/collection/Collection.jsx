import React from 'react';
import PropTypes from 'prop-types';
import indexedDB from 'libs/indexeddb/indexeddb';
import Header from 'ui/header/Header.jsx';
import InfoButton from './InfoButton.jsx';
import AddBooksButton from './AddBooksButton.jsx';
import CollectionBook from './CollectionBook.jsx';
import style from './Collection.scss';


/**
 * Collection view component
 */
class Collection extends React.Component {
  constructor(props) {
    super(props);

    this.onRemoveButtonClick = this.onRemoveButtonClick.bind(this);
    this.onInfoButtonClick = this.onInfoButtonClick.bind(this);
  }


  /**
   * Get collection from database if database is ready
   * @returns {undefined} undefined
   */
  componentDidMount() {
    const collectionState = this.props.collectionState;

    if (collectionState.dbReady) {
      this.props.getCollection();
    }

    this.props.updateCurrentBookData(null);
  }


  /**
   * Gets collection from database if the database wansn't ready when the
   * component mounted
   * @returns {undefined} undefined
   */
  componentDidUpdate(prevProps) {
    const prevState = prevProps.collectionState;
    const collectionState = this.props.collectionState;

    if (!prevState.dbReady &&
      collectionState.dbReady &&
      !collectionState.initialCollectionLoaded) {
      this.props.getCollection();
    }
  }


  /**
   * Click handler for the info button
   * @returns {undefined} undefined
   */
  onInfoButtonClick() {
    this.props.showInfo();
  }


  /**
   * Click handler for the book 'remove' buttons
   * @param {String} id - The id of the book to remove
   * @returns {undefined} undefined
   */
  onRemoveButtonClick(id) {
    const conf = confirm('Are you sure you want to delete this book?');

    if (conf) {
      indexedDB.deleteItem(id, 'books')
      .then(() => {
        this.props.getCollection();
      });
    }
  }


  /**
   * Renders message for an empty collection
   * @returns {ReactElement} Message for an empty collection
   */
  renderNoBooks() {
    const collectionState = this.props.collectionState;
    const loaderState = this.props.loaderState;
    let el;

    if (!loaderState.show && !collectionState.collection.length &&
      !collectionState.downloadingBook) {
      el = (
        <div className={style.message}>
          <p><span>You don't have any books in your collection.
          Click the button above right to add some from your <a href='https://dropbox.com' target='_blank' rel='noopener noreferrer'>Dropbox</a>.</span></p>
          <p>Don't have any ebooks? Download some for free at <a href='https://www.gutenberg.org/' target='_blank' rel='noopener noreferrer'>Project Gutenberg</a>.</p>
        </div>
      );
    }

    return el;
  }


  /**
   * Renders collection
   * @returns {Array<ReactElement>} An array of book buttons
   */
  renderBooks() {
    const collectionState = this.props.collectionState;
    const loaderState = this.props.loaderState;
    let books;

    if (!loaderState.show &&
      collectionState.collection.length &&
      !collectionState.downloadingBook) {
      books = [];

      for (let i = 0; i < collectionState.collection.length; i++) {
        const bookData = collectionState.collection[i];

        books.push(
          <CollectionBook
            data={bookData}
            key={'collection-book-' + i + '-' + bookData.id}
            removeButtonClickHandler={this.onRemoveButtonClick}
            windowSize={this.props.resizerState.windowSize}
          />,
        );
      }
    }

    return books;
  }


  /**
   * Renders component
   * @returns {ReactElement} The component
   */
  render() {
    const noBooksMessage = this.renderNoBooks();
    const books = this.renderBooks();

    return (
      <div className={style.wrapper}>
        <Header
          leftButton={
            <InfoButton clickHandler={this.onInfoButtonClick} />
          }
          rightButton={
            <AddBooksButton {...this.props} />
          }
          text={'Your books'}
        />
        <div className={style.booksWrapper}>
          {noBooksMessage}
          {books}
        </div>
      </div>
    );
  }
}


/**
 * PropTypes
 * @prop {Object} collectionState - The collection state in the store
 * @prop {Function} getCollection - function to dispatch a GET_COLLECTION action
 * @prop {Object} loaderState - The loader state in the store
 * @prop {Object} resizerState - The resizer state in the store
 * @prop {Function} showInfo - Function to dispatch a SHOW_INFO action
 * @prop {Function} updateCurrentBookData - Function to dispatch an
 * UPDATE_CURRENT_BOOK_DATA action
 */
Collection.propTypes = {
  collectionState: PropTypes.object.isRequired,
  getCollection: PropTypes.func.isRequired,
  loaderState: PropTypes.object.isRequired,
  resizerState: PropTypes.object.isRequired,
  showInfo: PropTypes.func.isRequired,
  updateCurrentBookData: PropTypes.func.isRequired,
};

export default Collection;
