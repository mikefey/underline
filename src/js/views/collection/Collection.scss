@import '~colors.scss';

.wrapper {
  float: left;
  height: 100%;
  position: relative;
  width: 100%;
}

.booksWrapper {
  float: left;
  height: 100vh;
  -webkit-overflow-scrolling: touch;
  overflow-y: scroll;
  padding: 60px 0 80px;
  position: relative;
  width: 100%;
}

.message {
  float: left;
  padding: 20px;
  text-align: left;
  width: 100%;

  p {
    &:first-child {
      margin-top: 0;
    }
  }

  a {
    color: $rose;
  }
}
