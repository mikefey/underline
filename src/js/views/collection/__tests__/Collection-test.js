import test from 'ava';
import React from 'react';
import { renderToString } from 'react-dom/server';

import Collection from './../Collection.jsx';

test('Collection component', (t) => {
  const readerState = {
    currentBookData: null,
    findingBookPosition: false,
  };
  const resizerState = {
    windowSize: {
      width: 1024,
      height: 768,
    },
  };
  const loaderState = {
    show: false,
  };
  const collectionState = {
    collection: [],
  };
  const downloadBook = () => {
    return true;
  };
  const getCollection = () => {
    return true;
  };
  const showInfo = () => {};
  const updateCurrentBookData = () => {};

  const tree = renderToString(<Collection
    loaderState={loaderState}
    collectionState={collectionState}
    downloadBook={downloadBook}
    getCollection={getCollection}
    readerState={readerState}
    resizerState={resizerState}
    showInfo={showInfo}
    updateCurrentBookData={updateCurrentBookData}
  />);

  t.snapshot(tree);
});
