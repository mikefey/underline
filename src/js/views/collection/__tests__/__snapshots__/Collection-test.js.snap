// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection component 1`] = `"<div class=\\"Collection__wrapper___Cqx9sa0pwqe_-pP\\" data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"837542035\\"><div class=\\"Header__wrapper___2PTLf_nkW7S2HEU\\" data-reactid=\\"2\\"><div class=\\"Header__content___2VIdu565Foup5Pb\\" data-reactid=\\"3\\"><button class=\\"InfoButton__wrapper___N_oMTzuA2GfeAJV\\" data-reactid=\\"4\\"><div class=\\"InfoButton__icon___1q8MRLZtz3Z3Myx\\" data-reactid=\\"5\\">i</div></button><h2 class=\\"Header__title___1_ebCiqkttbS44m\\" data-reactid=\\"6\\">Your books</h2><div class=\\"AddBooksButton__wrapper___3WnirUQZPEP4sp_\\" data-reactid=\\"7\\"><div data-reactid=\\"8\\"><div class=\\"AddBooksButton__icon___3Vo3Yz3c1Emxo_A\\" data-reactid=\\"9\\"></div></div></div></div></div><div class=\\"Collection__booksWrapper___JImf8enUFiESmgf\\" data-reactid=\\"10\\"><div class=\\"Collection__message___2JDyaCVrQufdN11\\" data-reactid=\\"11\\"><p data-reactid=\\"12\\"><span data-reactid=\\"13\\"><!-- react-text: 14 -->You don&#x27;t have any books in your collection. Click the button above right to add some from your <!-- /react-text --><a href=\\"https://dropbox.com\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\" data-reactid=\\"15\\">Dropbox</a><!-- react-text: 16 -->.<!-- /react-text --></span></p><p data-reactid=\\"17\\"><!-- react-text: 18 -->Don&#x27;t have any ebooks? Download some for free at <!-- /react-text --><a href=\\"https://www.gutenberg.org/\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\" data-reactid=\\"19\\">Project Gutenberg</a><!-- react-text: 20 -->.<!-- /react-text --></p></div></div></div>"`;
